KAFKA-16366: Refactor KTable source optimization

chickenchickenlove opened the pull request:
### Description

This PR refactors the `DSL`'s optimization logic (reusing the input topic as the changelog) to leverage the public `Topology#addReadOnlyStateStore` API introduced in KIP-813, replacing the previous usage of internal builder methods.

### Key Changes:
- Refactored `TableSourceNode`
  - Replaced the manual wiring of sources and stores (via `addSource`, `addProcessor`, `connectSourceStoreAndTopic`) with a call to `topologyBuilder.addReadOnlyStateStore`.
- Updated `InternalTopologyBuilder`:
  - Added an overloaded `addReadOnlyStateStore` method accepting a `ProcessorSupplier` with wildcard output types `(?, ?)`.
  - Reasoning: The public PAPI `addReadOnlyStateStore` enforces `Void, Void` output types. However, the `DSL`'s `KTableSource` needs to forward records downstream (emitting K, Change<V>). The internal overload allows the `DSL` to reuse the "read-only" wiring logic while maintaining its forwarding behavior.
- Added Regression Tests:
  - Added `TopologyTest#whenKTableSourceIsOptimizedThenTopologyShouldBeSerialPipeline` - Verifies that the generated topology description remains identical to the pre-refactoring structure, ensuring strict backward compatibility.
  - Added `TopologyTest#whenKTableSourceIsOptimizedThenItsStateStoreShouldNotLog` - Confirms that the optimization is correctly applied (changelog logging is disabled).

### Result
- Fixes https://issues.apache.org/jira/browse/KAFKA-16366

chickenchickenlove committed:
KAFKA-16366: Refactor KTable source optimization

KAFKA-16366: Refactor KTable source optimization

chickenchickenlove commented:
Hi @mjsax ! 
Just a gentle ping on this PR.
When you have some time, could you please take a look? Thank you. 🙇‍♂️

lucliu1108 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java:
```diff
@@ -58,6 +58,11 @@ public ProcessorParameters(final FixedKeyProcessorSupplier<KIn, VIn, VOut> proce
+        ApiUtils.checkSupplier(processorSupplier);
+        return topologyBuilder.wrapProcessorSupplier(processorName, processorSupplier);
+    }
 
```
Will this method be used outside the package? Might be good to make it package-private if only used in `TableSourceNode.java`

lucliu1108 replied:
The wrapping process is also repeated in `ProcessorParameters#addProcessorTo` method, maybe we want to substitute [these lines](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java#L68-L69) with the helper method?

This will make [processorSupplier block](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java#L67)  and [fixedKeyProcessorSupplier block](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java#L82) look not so parallel, so might be good to add a parallel `wrappedFixedKeyProcessSupplier` method
- it could be over-engineering on helper method, just some random thoughts

chickenchickenlove replied:
_so might be good to add a parallel wrappedFixedKeyProcessSupplier method_

It makes sense to me! 
By the way, since `wrappedFixedKeyProcessSupplier` doesn’t seem to be related to the `TableSource` optimization, how about handling it in a separate issue and PR to minimize the review scope?

Let me know your thoughts!

lucliu1108 replied:
Make sense to me! I agree this cleanup is beyond the scope of this ticket. We can leave it for now.

lucliu1108 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java:
```diff
@@ -96,27 +98,38 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
+
+                // TableSourceNode is backed by KTableSource, which maintains a single state store. 
+                // Therefore, we are guaranteed to have exactly one store.
+                assert tableSource.stores().size() == 1;
```
Could assert fail silently in production code(not completely sure) Looks like assert isn't used in other Java code.

Might be better to throw an exception instead?

lucliu1108 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java:
```diff
@@ -69,7 +74,7 @@ public void addProcessorTo(final InternalTopologyBuilder topologyBuilder, final
             final ProcessorSupplier<KIn, VIn, KOut, VOut> wrapped =
                 topologyBuilder.wrapProcessorSupplier(processorName, processorSupplier);
-
+            
```
Nit: extra spaces

lucliu1108 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java:
```diff
@@ -21,9 +21,11 @@
 import java.util.Collections;
 import java.util.Iterator;
+import java.util.Set;
 
```
Nit: These 2 imports added are unused.

lucliu1108 commented on streams/src/test/java/org/apache/kafka/streams/TopologyTest.java:
```diff
@@ -2438,6 +2438,64 @@ public void readOnlyStateStoresShouldNotLog() {
+        assertThat(stateStoreFactory.loggingEnabled(), equalTo(false));
+    }
+    
+
```
Not related to this test, but `TableSourceNodeTest#shouldConnectStateStoreToInputTopicIfInputTopicIsUsedAsChangelog` needs to be updated as well, since `connectSourceStoreAndTopic` is not explicitly called by topologyBuilder

chickenchickenlove replied:
Good point, I fixed it! Thanks a lot.

lucliu1108 reviewed (commented):
Hi @chickenchickenlove , sorry for getting back to you late, thanks for the PR! The refactoring looks good to me overall.

I left a few minor comments. Might be good for @mjsax to also take a look before modifying it.

chickenchickenlove commented:
@lucliu1108 
Thanks for taking look this PR!
Unfortunatelly, I'm on vacation without my laptop 🥲.
I will address my PR based on your comments as soon as I come back from vacation!

chickenchickenlove committed:
Addressing review and add more tests.

chickenchickenlove commented:
@lucliu1108 
Thank you for the detailed review, and I’m sorry for the late update!
I sincerely appreciate it.

I addressed your comment and added more tests. 
When you get a chance, please take another look. 🙇‍♂️

lucliu1108 reviewed (commented):
Hi @chickenchickenlove ! Sorry for getting back really late, and thanks for the revision, LGTM. 

Just one concern: 
In JIRA the goal is "refactor the DSL compilation code, to use this public API to build the `Topology` instead of internal APIs.", but we are still using the `InternalTopologyBuilder`'s own `addReadOnlyStateStore` (newly added) due to the fact that `Topology#addReadOnlyStateStore` constrains the supplier to `ProcessorSupplier<K, V, Void, Void>`. Personally I agree this is the workaround for the issue, and couldn't think of a cleaner way to call the public method directly without using some unchecked cast magic.

I'm wondering if a follow-up KIP to do some changes to `Topology#addReadOnlyStateStore` would make sense, for example adding an overload like:
```java
  public <K, V, KOut, VOut, S extends StateStore> Topology addReadOnlyStateStore(
      final StoreBuilder<S> storeBuilder,
      final String sourceName,
      ...,
      final ProcessorSupplier<K, V, KOut, VOut> stateUpdateSupplier
  )
```

Not entirely sure if it would worth it, and if there are other concerns associating with this additional flexibility

@mjsax could you advise on this when you get a chance? Thanks!
